tq/ssh.go

```diff
@@ -0,0 +1,391 @@
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	return conn.ReadStatusWithArguments()
+}
+
+// upload starts an upload.
+func (a *SSHAdapter) upload(t *Transfer, cb ProgressCallback) error {
+	rel, err := t.Rel("upload")
+	if err != nil {
+		return err
+	}
+	if rel == nil {
+		return errors.Errorf("No upload action for object: %s", t.Oid)
+	}
+
+	f, err := os.OpenFile(t.Path, os.O_RDONLY, 0644)
+	if err != nil {
+		return errors.Wrap(err, "basic upload")
+	}
+	defer f.Close()
+
+	status, _, text, err := a.doUpload(t, f, cb)
+	if err != nil {
+		return err
+	}
+	if status < 200 || status > 299 {
+		// A status code of 403 likely means that an authentication token for the
+		// upload has expired. This can be safely retried.
+		if status == 403 {
+			err = errors.New("http: received status 403")
+			return errors.NewRetriableError(err)
+		}
+
+		if status == 429 {
+			return errors.NewRetriableError(fmt.Errorf("got status %d when uploading OID %s", status, t.Oid))
+		}
+
+		if len(text) > 0 {
+			return fmt.Errorf("got status %d when uploading OID %s: %s", status, t.Oid, text[0])
+		}
+		return fmt.Errorf("got status %d when uploading OID %s", status, t.Oid)
+
+	}
+
+	return a.verifyUpload(t)
+}
+
+func (a *SSHAdapter) argumentsForTransfer(t *Transfer, action string) []string {
+	args := make([]string, 0, 3)
+	set, ok := t.Actions[action]
+	if !ok {
+		return nil
+	}
+	args = append(args, fmt.Sprintf("size=%d", t.Size))
+	if set.Id != "" {
+		args = append(args, fmt.Sprintf("id=%s", set.Id))
+	}
+	if set.Token != "" {
+		args = append(args, fmt.Sprintf("token=%s", set.Token))
+	}
+	return args
+}
+
+// Begin a new batch of uploads or downloads. Call this first, followed by
+// one or more Add calls. maxConcurrency controls the number of transfers
+// that may be done at once. The passed in callback will receive updates on
+// progress. Either argument may be nil if not required by the client.
+func (a *SSHAdapter) Begin(cfg AdapterConfig, cb ProgressCallback) error {
+	if err := a.adapterBase.Begin(cfg, cb); err != nil {
+		return err
+	}
+	a.ctx = a.adapterBase.apiClient.Context()
+	a.debugging = a.ctx.OSEnv().Bool("GIT_TRANSFER_TRACE", false)
+	return nil
+}
+
+// ClearTempStorage clears any temporary files, such as unfinished downloads that
+// would otherwise be resumed
+func (a *SSHAdapter) ClearTempStorage() error {
+	return nil
+}
+
+func (a *SSHAdapter) Trace(format string, args ...interface{}) {
+	if !a.adapterBase.debugging {
+		return
+	}
+	tracerx.Printf(format, args...)
+}
+
+func configureSSHAdapter(m *Manifest) {
+	m.RegisterNewAdapterFunc("ssh", Upload, func(name string, dir Direction) Adapter {
+		bu := &SSHAdapter{newAdapterBase(m.fs, name, dir, nil), nil, m.sshTransfer}
+		bu.transferImpl = bu
+		return bu
+	})
+	m.RegisterNewAdapterFunc("ssh", Download, func(name string, dir Direction) Adapter {
+		bu := &SSHAdapter{newAdapterBase(m.fs, name, dir, nil), nil, m.sshTransfer}
+		bu.transferImpl = bu
+		return bu
```

Comment:
Yeah, I can change the variable name.  I intentionally didn't replicate the `panic` call because it doesn't matter: the adapter is the same and changing the direction doesn't change the adapter.  For our upload and download adapters, it does very much matter indeed, but not here.